src/utils/untar.js

```diff
@@ -0,0 +1,316 @@
+    // An extended header shall consist of one or more records, each constructed as follows:
+    // "%d %s=%s\n", <length>, <keyword>, <value>
+
+    // The extended header records shall be encoded according to the ISO/IEC10646-1:2000 standard (UTF-8).
+    // The <length> field, <blank>, equals sign, and <newline> shown shall be limited to the portable character set, as
+    // encoded in UTF-8. The <keyword> and <value> fields can be any UTF-8 characters. The <length> field shall be the
+    // decimal length of the extended header record in octets, including the trailing <newline>.
+
+    let bytes = new Uint8Array(buffer);
+    const fields = [];
+
+    const textDecoder = new TextDecoder();
+
+    while (bytes.length > 0) {
+      // Decode bytes up to the first space character; that is the total field length
+      const fieldLength = parseInt(
+        textDecoder.decode(bytes.subarray(0, bytes.indexOf(0x20))),
+        10
+      );
+      const fieldText = textDecoder.decode(bytes.subarray(0, fieldLength));
+      const fieldMatch = fieldText.match(/^\d+ ([^=]+)=(.*)\n$/);
+
+      if (fieldMatch === null) {
+        throw new Error('Invalid PAX header data format.');
+      }
+
+      const fieldName = fieldMatch[1];
+      let fieldValue = fieldMatch[2];
+
+      if (fieldValue.length === 0) {
+        fieldValue = null;
+      } else if (fieldValue.match(/^\d+$/) !== null) {
+        // If it's a integer field, parse it as int
+        fieldValue = parseInt(fieldValue, 10);
+      }
+      // Don't parse float values since precision is lost
+
+      const field = {
+        name: fieldName,
+        value: fieldValue,
+      };
+
+      fields.push(field);
+
+      bytes = bytes.subarray(fieldLength); // Cut off the parsed field data
+    }
+
+    return new PaxHeader(fields);
+  }
+
+  constructor(fields: FieldEntry[]) {
+    this._fields = fields;
+  }
+
+  applyHeader(file) {
+    // Apply fields to the file
+    // If a field is of value null, it should be deleted from the file
+    // https://www.mkssoftware.com/docs/man4/pax.4.asp
+
+    this._fields.forEach(function (field) {
+      let fieldName = field.name;
+      const fieldValue = field.value;
+
+      if (fieldName === 'path') {
+        // This overrides the name and prefix fields in the following header block.
+        fieldName = 'name';
+
+        if (file.namePrefix !== undefined) {
+          delete file.namePrefix;
+        }
+      } else if (fieldName === 'linkpath') {
+        // This overrides the linkname field in the following header block.
+        fieldName = 'linkname';
+      }
+
+      if (fieldValue === null) {
+        delete file[fieldName];
+      } else {
+        file[fieldName] = (fieldValue: any);
+      }
+    });
+  }
+}
+
+// TODO: Consider unifying this ByteReader with the one in art-trace.js
+class ByteReader {
+  _uint8Array: Uint8Array;
+  _bufferView: DataView;
+  _position = 0;
+  _textDecoder = new TextDecoder();
```

Comment:
being pedantic: all the values in the header are ascii and not UTF-8 -- which should work for well-formed tar files, because of course ascii is part of UTF-8. But could there be a risk from malformed tar files?

I believe a simple fix is that we can use `'ascii'` in the constructor? (this maps to windows-1252 according to https://developer.mozilla.org/en-US/docs/Web/API/Encoding_API/Encodings -- still not 100% correct but could be less garbled in case of encoding problems?).

I believe only pax header values use UTF-8.

Reply:
I'm giving this a try.